### ISSUE TITLE
Incorrect Derivative of LogSoftMax layer

### ISSUE THREAD

**Issue by mrdaybird** [mlpack/mlpack#3468]
**Description**: The derivative used in the Backward method of the LogSoftMax layer is incorrect. This issue seemed to be noticed in the past but was closed, possibly incorrectly(?) (See #2364  )

The derivative calculated using LogSoftMax should be equal to derivative of logarithm applied on the Softmax layer. But, the derivative using both are significantly different. 

**Code to reproduce**
```
#include <mlpack.hpp>

using namespace mlpack;
using namespace arma;

int main(){
  int n = 5, m = 1;
  mat input = arma::mat("-0.6871; 0.7898; 0.2011; 0.0949; -0.0550");
  input.print("Input");
  mat output(n, m), delta(n, m); 

  LogSoftMax lsm;
  lsm.Forward(input, output);
  mat gy = ones(n, m); 
  lsm.Backward(output, gy, delta);

  mat outputT(n, m), delta2(n, m); 

  Softmax sm; 
  sm.Forward(input, outputT);
  mat output2 = arma::log(outputT);
  //Derivative of log function. 
  gy = arma::pow(outputT, -1);
  
  //Derivative of the Softmax layer 
  sm.Backward(outputT, gy, delta2);

  output.print("LogSoftMax Output:");
  output2.print("Log on Softmax Output:");

  delta.print("LogSoftMax Derivative:");
  delta2.print("Log on SoftMax Derivative:");

  return 0;
}
```
Output of the above code: 

```
Input
  -0.6871
   0.7898
   0.2011
   0.0949
  -0.0550
LogSoftMax Output:
  -2.4746
  -0.9977
  -1.5864
  -1.6926
  -1.8425
Log on Softmax Output:
  -2.4746
  -0.9977
  -1.5864
  -1.6926
  -1.8425
LogSoftMax Derivative:
   1.0842
   1.3687
   1.2047
   1.1840
   1.1584
Log on SoftMax Derivative:
   0.5790
  -0.8435
  -0.0233
   0.0798
   0.2079

```

Also, derivative of Softmax was verified manually and the derivative calculated using "Log on Softmax" is also verified using pytorch using the following code: 

```
import torch

input = torch.tensor([-0.6871, 0.7898, 0.2011, 0.0949,-0.0550], requires_grad=True)
output = torch.nn.LogSoftmax(dim=0)(input)
obj = output.sum()
output.retain_grad()
obj.backward()
print("Derivative of the input:", input.grad)
```

Output:

`Derivative of the input: tensor([ 0.5790, -0.8435, -0.0233,  0.0798,  0.2079])`
